rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,430 @@
+requiring them may make the life of our other compiler backends harder.
+
+As mentioned above, exposing more than loads and stores, and non-`Relaxed`
+atomic memory orderings, also muddies the "a volatile op should compile into one
+hardware memory instruction" narrative that is so convenient for loads and stores.
+Further, compatibility of non-load/store atomics in IPC scenario may require
+some ABI agreement on how atomics should be implemented between the interacting
+programs.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+I think we want to do this because it simultaneously eliminates two well-known
+and unnecessary volatile footguns (data races and tearing) and opens interesting
+new possibilities in interprocess communication.
+
+But although this feature may seem simple, its design space is actually
+remarquably large, and a great many alternatives were considered before reaching
+the current design proposal. Here is a map of some design knobs that I explored:
+
+## Extending `AtomicXyz`
+[extending-atomicxyz]: #extending-atomicxyz
+
+Atomic volatile operations could be made part of the Atomic wrapper types.
+However, they cannot be just another atomic memory ordering, due to the fact
+that atomic ops take an `&self` (which is against the spirit of volatile as it
+is tagged with `dereferenceable` at the LLVM layer).
+
+Instead, one would need to add more methods to atomic wrapper types, which take
+a pointer as a self-parameter. I thought that this API inconsistency would be
+jarring to users, not to mention that potentially having a copy of each atomic
+memory operation with a `_volatile` prefix would get annoying quickly when
+reading through the API docs.
+
+Also, it is extremely common to want either all operations on a memory
+location to be volatile, or none of them. Providing separate wrapper types
+helps enforce this very common usage pattern at the API level.
+
+## Self-type or not self-type
+[self-type]: #self-type
+
+Instead of using `arbitrary_self_types` to get `a.load(Ordering::Relaxed)`
+method syntax on pointer-like objects, one could instead provide the volatile
+operations as inherent methods, i.e. `VolatileU8::load(a, Ordering::Relaxed)`.
+
+This has the advantage of avoiding coupling this feature to another unstable
+feature. But it has the drawback of being incredibly verbose. Typing the same
+volatile type name over and over again in a complex shared memory transaction
+would certainly get old and turn annoying quickly, and we don't want anger to
+distract low-level developers from the delicate task of implementing the kind
+of subtle algorithm that requires volatile operations.
+
+## `NonNull<T>` vs `*mut T` vs `*const T` vs other
+[pointers]: #pointers
+
+Honestly, I don't have a very strong opinion there. I have a positive _a priori_
+towards `NonNull<T>` because it encodes an important invariant in the API, and
```

Comment:
The invariant might be worth re-stating here (e.g. volatile reads / writes assume that the pointer is non-null, aligned, and dereferenceable), and which parts of this invariant are encoded by the `NonNull` API.

Reply:
I'll do that. Another thing which I should perhaps do is to remove the first-person discourse, upon re-reading it it seems to fit neither the rest of the text nor the overall RFC format.